gobblin-yarn/src/main/java/gobblin/yarn/GobblinYarnAppLauncher.java

```diff
@@ -0,0 +1,598 @@
+        System.getenv(ApplicationConstants.Environment.JAVA_HOME.key()));
+
+    // Add jars/files in the working directory of the ApplicationMaster to the CLASSPATH
+    Apps.addToEnvironment(environmentVariableMap, ApplicationConstants.Environment.CLASSPATH.key(),
+        ApplicationConstants.Environment.PWD.$());
+    Apps.addToEnvironment(environmentVariableMap, ApplicationConstants.Environment.CLASSPATH.key(),
+        ApplicationConstants.Environment.PWD.$() + File.separator + "*");
+
+    String[] classpaths = this.yarnConfiguration.getStrings(YarnConfiguration.YARN_APPLICATION_CLASSPATH,
+        YarnConfiguration.DEFAULT_YARN_APPLICATION_CLASSPATH);
+    if (classpaths != null) {
+      for (String classpath : classpaths) {
+        Apps.addToEnvironment(
+            environmentVariableMap, ApplicationConstants.Environment.CLASSPATH.key(), classpath.trim());
+      }
+    }
+
+    return environmentVariableMap;
+  }
+
+  private String buildApplicationMasterCommand(int memoryMbs) {
+    String appMasterClassName = GobblinApplicationMaster.class.getSimpleName();
+    return String.format(
+        "%s/bin/java -Xmx%dM %s %s" +
+        " --%s %s" +
+        " 1>" + ApplicationConstants.LOG_DIR_EXPANSION_VAR + File.separator + "%s.%s" +
+        " 2>" + ApplicationConstants.LOG_DIR_EXPANSION_VAR + File.separator + "%s.%s",
+        ApplicationConstants.Environment.JAVA_HOME.$(), memoryMbs, this.appMasterJvmArgs,
+        GobblinApplicationMaster.class.getName(), ConfigurationConstants.APPLICATION_NAME_OPTION_NAME,
+        this.appName, appMasterClassName, ApplicationConstants.STDOUT, appMasterClassName,
+        ApplicationConstants.STDERR);
+  }
+
+  private void setupSecurityTokens(ContainerLaunchContext containerLaunchContext) throws IOException {
+    Credentials credentials = UserGroupInformation.getCurrentUser().getCredentials();
+    String tokenRenewer = this.yarnConfiguration.get(YarnConfiguration.RM_PRINCIPAL);
+    if (tokenRenewer == null || tokenRenewer.length() == 0) {
+      throw new IOException("Failed to get master Kerberos principal for the RM to use as renewer");
+    }
+
+    // For now, only getting tokens for the default file-system.
+    Token<?> tokens[] = this.fs.addDelegationTokens(tokenRenewer, credentials);
+    if (tokens != null) {
+      for (Token<?> token : tokens) {
+        LOGGER.info("Got delegation token for " + this.fs.getUri() + "; " + token);
+      }
+    }
+
+    Closer closer = Closer.create();
+    try {
+      DataOutputBuffer dataOutputBuffer = closer.register(new DataOutputBuffer());
+      credentials.writeTokenStorageToStream(dataOutputBuffer);
+      ByteBuffer fsTokens = ByteBuffer.wrap(dataOutputBuffer.getData(), 0, dataOutputBuffer.getLength());
+      containerLaunchContext.setTokens(fsTokens);
+    } catch (Throwable t) {
+      throw closer.rethrow(t);
+    } finally {
+      closer.close();
+    }
+  }
+
+  private void sendShutdownRequest() {
```

Comment:
I noticed that sometime `killApplication` does not properly kill the containers, and that's why I resorted to this approach for more graceful shutdown.